tests/withBlockTracker.ts

```diff
@@ -0,0 +1,252 @@
+ * provider.
+ * @param callback - A callback which will be called with the built block
+ * tracker.
+ */
+async function withPollingBlockTracker(
+  options: WithPollingBlockTrackerTypes['withBlockTrackerOptions'],
+  callback: WithPollingBlockTrackerTypes['withBlockTrackerCallback'],
+): Promise<void>;
+/**
+ * Calls the given function with a built-in PollingBlockTracker, ensuring that
+ * all listeners that are on the block tracker are removed and any timers or
+ * loops that are running within the block tracker are properly stopped.
+ *
+ * @param callback - A callback which will be called with the built block
+ * tracker.
+ */
+async function withPollingBlockTracker(
+  callback: WithPollingBlockTrackerTypes['withBlockTrackerCallback'],
+): Promise<void>;
+/* eslint-disable-next-line jsdoc/require-jsdoc */
+async function withPollingBlockTracker(...args: any[]) {
+  const callback: WithPollingBlockTrackerTypes['withBlockTrackerCallback'] =
+    args.pop();
+  const options =
+    (args[0] as WithPollingBlockTrackerTypes['withBlockTrackerOptions']) ?? {};
+  const provider =
+    options.provider === undefined
+      ? new FakeProvider()
+      : new FakeProvider(options.provider);
+
+  const blockTrackerOptions =
+    options.blockTracker === undefined
+      ? { provider }
+      : {
+          provider,
+          ...options.blockTracker,
+        };
+  const blockTracker = new PollingBlockTracker(blockTrackerOptions);
+  const callbackArgs = { provider, blockTracker };
+  await callback(callbackArgs);
+}
+
+/**
+ * Calls the given function with a built-in SubscribeBlockTracker, ensuring that
+ * all listeners that are on the block tracker are removed and any timers or
+ * loops that are running within the block tracker are properly stopped.
+ *
+ * @param options - Options that allow configuring the block tracker or
+ * provider.
+ * @param callback - A callback which will be called with the built block
+ * tracker.
+ */
+async function withSubscribeBlockTracker(
+  options: WithSubscribeBlockTrackerTypes['withBlockTrackerOptions'],
+  callback: WithSubscribeBlockTrackerTypes['withBlockTrackerCallback'],
+): Promise<void>;
+/**
+ * Calls the given function with a built-in SubscribeBlockTracker, ensuring that
+ * all listeners that are on the block tracker are removed and any timers or
+ * loops that are running within the block tracker are properly stopped.
+ *
+ * @param callback - A callback which will be called with the built block
+ * tracker.
+ */
+async function withSubscribeBlockTracker(
+  callback: WithSubscribeBlockTrackerTypes['withBlockTrackerCallback'],
+): Promise<void>;
+/* eslint-disable-next-line jsdoc/require-jsdoc */
+async function withSubscribeBlockTracker(...args: any[]): Promise<void> {
+  const callback: WithSubscribeBlockTrackerTypes['withBlockTrackerCallback'] =
+    args.pop();
+  const options =
+    (args[0] as WithSubscribeBlockTrackerTypes['withBlockTrackerOptions']) ??
+    {};
```

Comment:
We could avoid using `as` here by including the full argument types. It doesn't seem to interfere with the overrides being picked up by VSCode.

e.g.
```suggestion
async function withPollingBlockTracker(
  ...args:
    | [
        WithPollingBlockTrackerTypes['withBlockTrackerOptions'],
        WithPollingBlockTrackerTypes['withBlockTrackerCallback'],
      ]
    | [WithPollingBlockTrackerTypes['withBlockTrackerCallback']]
) {
  let callback: WithPollingBlockTrackerTypes['withBlockTrackerCallback'];
  let options: WithPollingBlockTrackerTypes['withBlockTrackerOptions'] = {};

  if (args.length === 2) {
    options = args[0];
    callback = args[1];
  } else {
    callback = args[0];
  }
```